src/modules/filesystem/NativeFile.cpp

```diff
@@ -311,6 +330,75 @@ File::Mode NativeFile::getMode() const
 	return mode;
 }
 
+bool NativeFile::setupBuffering(BufferMode mode, int64 bufferSize)
+{
+	int8 *newbuf = nullptr;
+	if (mode != BUFFER_NONE)
+	{
+		newbuf = new (std::nothrow) int8[(size_t) bufferSize];
+		if (newbuf == nullptr)
+			return false;
+	}
+
+	delete[] buffer;
+	buffer = newbuf;
+	bufferUsed = this->mode == MODE_READ ? bufferSize : 0;
+	return true;
+}
+
+int64 NativeFile::bufferedRead(void *dst, int64 size)
+{
+	int8 *ptr = (int8 *) dst;
+	int64 readed = 0;
+
+	while (size > 0)
+	{
+		int64 available = bufferSize - bufferUsed;
+
+		if (available > 0)
+		{
+			// There's leftover buffers.
+			size_t copy = (size_t) std::min(size, available);
+			memcpy(ptr, buffer + (size_t) bufferUsed, copy);
+
+			ptr += copy;
+			size -= (int64) copy;
+			bufferUsed += (int64) copy;
+			readed += copy;
+		}
+		else
+		{
+			// Buffer is empty. Fill it.
+			size_t ureaded = SDL_ReadIO(file, buffer, (size_t) bufferSize);
+			bufferUsed = bufferSize - (int64) ureaded;
+
+			if (ureaded == 0)
+				break;
+
+			if (bufferUsed > 0)
+				// Shift the buffer so code above can properly index it
+				memmove(buffer + bufferUsed, buffer, ureaded);
+		}
+	}
+
+	return readed;
+}
+
+bool NativeFile::bufferedWrite(const void *data, int64 size)
+{
+	if ((size + bufferUsed) < bufferSize)
+	{
+		// Entire data fits in the buffer.
+		memcpy(buffer + bufferUsed, data, size);
+		bufferUsed += size;
+		return true;
+	}
+
+	// Could overflow. Write directly.
+	size_t writeSize = (size_t) size;
+	return flush() && (SDL_WriteIO(file, data, writeSize) == writeSize);
```

Comment:
Is a full flush including `SDL_FlushIO` needed here? Or could we make it do most of the work of `flush()` but skip `SDL_FlushIO`?

This will also end up calling `SDL_FlushIO` before `SDL_WriteIO` even if no data is in the buffer, I think.

Reply:
That one is kinda lazy. I implemented more sophisticated buffered write routines instead.